docs: Add description of experimental and regular datasets to contribution guide (#642)
kedro-org/kedro-plugins: kedro-datasets/CONTRIBUTING.md

If you have new ideas for Kedro-Datasets then please open a [GitHub issue](https://github.com/kedro-org/kedro-plugins/issues) with the label `enhancement`. Please describe in your own words the feature you would like to see, why you need it, and how it should work.

## Contribute a new dataset

If you're unsure where to begin contributing to Kedro-Datasets, please start by looking through the `good first issue` and `help wanted` on [GitHub](https://github.com/kedro-org/kedro-plugins/issues).
If you want to contribute a new dataset, read the [tutorial to create and contribute a custom dataset](https://docs.kedro.org/en/stable/data/how_to_create_a_custom_dataset.html) in the Kedro documentation.
Make sure to add the new dataset to `kedro_datasets.rst` so that it shows up in the API documentation and to `static/jsonschema/kedro-catalog-X.json` for IDE validation.

> Below is a guide to help you understand the process of contributing a new dataset, whether it falls under the category of regular or experimental datasets.

Author: Done!

### Difference between regular and experimental datasets

> #### Regular datasets

Reviewer: ```suggestion
#### 1st party supported datasets
```

I think regular is a bit too vague, something like this is really explicit from the title

Author: I've just changed this to "Core" after @deepyaman's suggestion. Does that sound better?

Reviewer: Yeah much better!

Regular datasets are maintained by the [Kedro Technical Steering Committee (TSC)](https://docs.kedro.org/en/stable/contribution/technical_steering_committee.html) and adhere to specific standards. These datasets adhere to the following requirements:

1. Datasets that the Kedro TSC is willing to maintain.
2. Must be fully documented.
3. Must have working doctests (unless complex cloud/DB setup required, which can be discussed in the review).
4. Must run as part of the regular CI/CD jobs.
5. Must have 100% test coverage.
6. Should support all Python versions under NEP 29 (3.9+ currently).

> 7. Should work on Ubuntu, Linux, and Windows.

Reviewer 2: ```suggestion
7. Should work on Linux, macOS, and Windows.
```

?

> #### Experimental Datasets

Reviewer 2: ```suggestion
#### Experimental datasets
```

> The requirements for experimental datasets are more flexible and these datasets are not maintained by the Kedro TSC. Experimental datasets:

Reviewer: Should we provide an example - e.g. a new DataFrame framework which hasn't yet seen mass adoption or a niche Simulation format.

Author: I thought about that, but if the TSC is willing to maintain it even something like that could become a "core" contribution.

> 1. Do not need to be fully documented but must have doc strings explaining their use.

Reviewer 2: ```suggestion
1. Do not need to be fully documented but must have docstrings explaining their use.
```

> 2. Do not need to run as part of regular CI/CD jobs.

Author: But the experimental datasets don't need to be maintained like the core ones. Obviously if an experimental dataset doesn't work at all and users complain, we can just remove it altogether. But IMO they serve also as inspiration for others on creating their own datasets and might not immediately work out of the box.

> 3. Can be in the early stages of development or do not have to meet the criteria for regular Kedro datasets.

Author: Dependency resolution is a topic we will cover separately. The model we currently have doesn't work either, so this isn't a unique problem for experimental datasets.

### Graduation of datasets
If your dataset is initially considered experimental but matures over time, it may qualify for graduation to a regular dataset.

> 1. Anyone, including TSC members and users, can trigger the graduation process.

Author: Aren't the lists above enough? I think we should just give this process a try and then we can always go and clarify things where needed.

> 2. An experimental dataset requires 1/2 approval from the TSC to graduate to the regular datasets space.

Author: > What is the 1/2 approval for? Is it basically just answering whether "the Kedro TSC is willing to maintain" the dataset, from the core criteria above?

Yes

Copying a comment over from #583 "Yes I think it might be possible [to use CODEOWNERS], but after reading this https://www.arnica.io/blog/what-every-developer-should-know-about-github-codeowners I wasn't so sure if CODEOWNERS is really the way we want to go?"

3. Your dataset can graduate when it meets all requirements of a regular dataset.

### Demotion of datasets

> A dataset initially considered regular might be demoted if it no longer meets the required standards.

Reviewer 3: Do we want to mention if the 'dataset' has low/no usage. I reckon 1. 'Datasets that the Kedro TSC is willing to maintain' vaguely covers that?

Author: How do we know if a dataset has low/no usage? We can't currently track that with telemetry..

Reviewer 2: Indeed, we do not. We will need to trust our own judgement on this.

| `fix`            | Non-breaking change which fixes an issue                                    |
| `tests`          | Changes to project unit (`tests/`) and / or integration (`features/`) tests |

## Dataset contribution process

 1. Fork the project
 2. Develop your contribution in a new branch.

>  3. Add your dataset to the appropriate location:

Reviewer 2: What's not clear about this workflow is how first-time contributors can know ahead of time whether the TSC will be willing to maintain the dataset, and therefore where to contribute it.

Author: You're totally right. Should we just say they open it as `experimental` and then the TSC does a vote on whether we want to maintain it or not? Which is basically a vote for "immediate graduation".

>     - `kedro-datasets/experimental` for experimental datasets.

Reviewer 2: Reading this made me realize that I'm not 100 % sure whether we decided to _ship_ these datasets as part of the `kedro-datasets` PyPI package or not. If we said "yes", then maybe we need to tweak either our packaging or the location of this directory. If we said "no", then it would be good to clarify it in this guide and https://github.com/kedro-org/kedro-plugins/blob/main/kedro-datasets/experimental/README.md.

Author: Good point! I think we should ship them to make it as easy as possible to use them e.g. just import `from kedro_datasets import experimental`.

Author: The results of the poll was a unanimous vote from 11 TSC members to include experimental contributions in the `kedro-datasets` package. I'll create a follow up task to update our packaging.

Reviewer 4: > Good point! I think we should ship them to make it as easy as possible to use them e.g. just import `from kedro_datasets import experimental`.

I think it's fine to even have them be importable the same way/top-level `kedro_datasets` namespace. As in, your catalog entry doesn't have to be `experimental.sap.HANADataset`. Although I'm not 100% sure on this, maybe there is value to being explicit to the reader that they're using something coming from `experimental`.

    - `kedro-datasets/kedro_datasets` for datasets meeting regular standards.
 4. Make sure all your commits are signed off by using `-s` flag with `git commit`.

>  5. Open a PR against the `main` branch and sure that the PR title follows the [Conventional Commits specs](https://www.conventionalcommits.org/en/v1.0.0/) with the scope `(datasets)`.

Reviewer 5: I think the word "make" is missing here? 

"Open a PR against the `main` branch and **make** sure that..."

 6. The TSC will review your contribution and decide whether it fits as a regular or experimental dataset.

>  7. Make sure the CI builds are green (have a look at the section [Running checks locally](#running-checks-locally) below)
>  8. Update the PR according to the reviewer's comments

Reviewer 4: ```suggestion
 7. Make sure the CI builds are green (have a look at the section [Running checks locally](#running-checks-locally) below).
 8. Update the PR according to the reviewer's comments.
```
. :)
